File: app/src/main/java/com/github/shadowsocks/utils/IOUtils.java
package com.github.shadowsocks.utils;
/*
 * Shadowsocks - A shadowsocks client for Android
 * Copyright (C) 2014 <max.c.lv@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 *                            ___====-_  _-====___
 *                      _--^^^#####//      \\#####^^^--_
 *                   _-^##########// (    ) \\##########^-_
 *                  -############//  |\^^/|  \\############-
 *                _/############//   (@::@)   \\############\_
 *               /#############((     \\//     ))#############\
 *              -###############\\    (oo)    //###############-
 *             -#################\\  / VV \  //#################-
 *            -###################\\/      \//###################-
 *           _#/|##########/\######(   /\   )######/\##########|\#_
 *           |/ |#/\#/\#/\/  \#/\##\  |  |  /##/\#/  \/\#/\#/\#| \|
 *           `  |/  V  V  `   V  \#\| |  | |/#/  V   '  V  V  \|  '
 *              `   `  `      `   / | |  | | \   '      '  '   '
 *                               (  | |  | |  )
 *                              __\ | |  | | /__
 *                             (vvv(VVV)(VVV)vvv)
 *
 *                              HERE BE DRAGONS
 *
 */

import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Reader;
import java.io.Writer;
import java.net.HttpURLConnection;

public class IOUtils {

    private static final String TAG = IOUtils.class.getSimpleName();
    private static final int BUFFER_SIZE = 32 * 1024;

    /**
     * in copy to out
     *
     * @param in  input stream
     * @param out output stream
     * @throws IOException io write error
     */
    public static void copy(InputStream in, OutputStream out) throws IOException {
        byte[] buffer = new byte[BUFFER_SIZE];
        int temp;
        while ((temp = in.read(buffer)) != -1) {
            out.write(buffer, 0, temp);
        }
        out.flush();
    }

    /**
     * read string by input stream
     *
     * @param in input stream
     * @return read failed return ""
     * @throws IOException io write error
     */
    public static String readString(InputStream in) throws IOException {
        StringBuilder builder = new StringBuilder();
        byte[] buffer = new byte[BUFFER_SIZE];
        int temp;
        while ((temp = in.read(buffer)) != -1) {
            builder.append(new String(buffer, 0, temp));
        }
        return builder.toString();
    }

    /**
     * write string
     *
     * @param file    file path
     * @param content string content
     * @return write failed return false.
     */
    public static boolean writeString(String file, String content) {
        FileWriter writer = null;
        try {
            writer = new FileWriter(file);
            writer.write(content);
            return true;
        } catch (IOException e) {
            VayLog.e(TAG, "writeString", e);
            return false;
        } finally {
            try {
                if (writer != null) {
                    writer.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public static void close(InputStream inputStream) {
        try {
            if (inputStream != null) {
                inputStream.close();
            }
        } catch (IOException e) {
            // ignored
        }
    }

    public static void close(OutputStream outputStream) {
        try {
            if (outputStream != null) {
                outputStream.close();
            }
        } catch (IOException e) {
            // Ignored
        }
    }

    public static void close(Writer writer) {
        try {
            if (writer != null) {
                writer.close();
            }
        } catch (IOException e) {
            // Ignored
        }
    }

    public static void close(Reader reader) {
        try {
            if (reader != null) {
                reader.close();
            }
        } catch (IOException e) {
            // Ignored
        }
    }

    public static void disconnect(HttpURLConnection conn) {
        if (conn != null) {
            conn.disconnect();
        }
    }
}

File: app/src/main/java/com/bige0/shadowsocksr/utils/IOUtils.kt
package com.bige0.shadowsocksr.utils

import java.io.*
import java.net.*

object IOUtils
{
	private const val TAG = "IOUtils"
	private const val BUFFER_SIZE = 32 * 1024

	/**
	 * inputStream copy to out
	 *
	 * @param inputStream  input stream
	 * @param out output stream
	 */
	fun copy(inputStream: InputStream, out: OutputStream)
	{
		val buffer = ByteArray(BUFFER_SIZE)
		var temp: Int
		while (true)
		{
			temp = inputStream.read(buffer)
			if (temp == -1)
			{
				break
			}
			out.write(buffer, 0, temp)
		}
		out.flush()
	}

	/**
	 * read string by input stream
	 *
	 * @param `inputStream` input stream
	 * @return read failed return ""
	 */
	fun readString(inputStream: InputStream): String
	{
		val builder = StringBuilder()
		val buffer = ByteArray(BUFFER_SIZE)
		var temp: Int
		while (true)
		{
			temp = inputStream.read(buffer)
			if (temp == -1)
			{
				break
			}
			builder.append(String(buffer, 0, temp))
		}
		return builder.toString()
	}

	/**
	 * write string
	 *
	 * @param file    file path
	 * @param content string content
	 * @return write failed return false.
	 */
	fun writeString(file: String, content: String): Boolean
	{
		var writer: FileWriter? = null
		try
		{
			writer = FileWriter(file)
			writer.write(content)
			return true
		}
		catch (e: IOException)
		{
			VayLog.e(TAG, "writeString", e)
			return false
		}
		finally
		{
			try
			{
				writer?.close()
			}
			catch (e: IOException)
			{
				e.printStackTrace()
			}
		}
	}

	fun close(inputStream: InputStream?)
	{
		try
		{
			inputStream?.close()
		}
		catch (e: IOException)
		{
			// ignored
		}
	}

	fun close(outputStream: OutputStream?)
	{
		try
		{
			outputStream?.close()
		}
		catch (e: IOException)
		{
			// Ignored
		}
	}

	fun close(writer: Writer?)
	{
		try
		{
			writer?.close()
		}
		catch (e: IOException)
		{
			// Ignored
		}
	}

	fun close(reader: Reader?)
	{
		try
		{
			reader?.close()
		}
		catch (e: IOException)
		{
			// Ignored
		}
	}

	fun disconnect(conn: HttpURLConnection?)
	{
		conn?.disconnect()
	}
}
